src/function/gds/output_writer.cpp

```diff
@@ -104,99 +107,139 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
+    // are enumerating the parents of the destination, then we should switch to
+    // forward-tracking if the next parent has visited the destination at a length
+    // that's greater than or equal to the lower bound of the recursive join. Otherwise,
+    // we'll enumerate paths that are smaller than the lower; OR
+    if (path.size() == 1 && nextIter >= info.lowerBound) {
         return true;
     }
-    // Path node predicate should only be applied to intermediate node. So we exclude dst node.
-    for (auto i = 0u; i < path.size() - 1; ++i) {
-        if (!info.pathNodeMask->valid(path[i]->getNodeID())) {
-            return false;
-        }
+    // (2) if this is not the first element in the stack, i.e., then we should switch
+    // to forward tracking only if the next parent of the top node in the stack has the
+    // same iter value as the current parent. That's because the levels/iter need to
+    // decrease by 1 each time we add a new node in the stack.
+    if (nextIter == getTop(path)->getIter()) {
+        return true;
     }
-    return true;
+    return false;
+}
+
+bool PathsOutputWriter::checkNewPathElement(const std::vector<ParentList*>& path,
+    ParentList* element) const {
+    return checkPathNodeMask(element) && checkSemantic(path, element);
 }
 
-// TODO(Xiyang): apply semantic check during DFS backtracking.
-bool PathsOutputWriter::checkSemantic(const std::vector<ParentList*>& path) const {
+bool PathsOutputWriter::checkPathNodeMask(ParentList* element) const {
+    if (!info.hasNodeMask() || element->getIter() == 1) {
+        return true;
+    }
+    return info.pathNodeMask->valid(element->getNodeID());
+}
+
+bool PathsOutputWriter::checkSemantic(const std::vector<ParentList*>& path,
+    ParentList* candidate) const {
     switch (info.semantic) {
     case PathSemantic::WALK:
         return true;
     case PathSemantic::TRAIL:
-        return isTrail(path);
+        return isTrail(path, candidate);
     case PathSemantic::ACYCLIC:
-        return isAcyclic(path);
+        return isAcyclic(path, candidate);
     default:
         KU_UNREACHABLE;
     }
 }
 
-bool PathsOutputWriter::isTrail(const std::vector<ParentList*>& path) const {
-    rel_id_set_t set;
-    for (auto i = 0u; i < path.size(); ++i) {
-        auto edgeID = path[i]->getEdgeID();
-        if (set.contains(edgeID)) {
+bool PathsOutputWriter::isTrail(const std::vector<ParentList*>& path, ParentList* candidate) const {
+    for (auto& element : path) {
+        if (candidate->getEdgeID() == element->getEdgeID()) {
             return false;
         }
-        set.insert(edgeID);
     }
     return true;
 }
 
-bool PathsOutputWriter::isAcyclic(const std::vector<ParentList*>& path) const {
-    node_id_set_t set;
-    for (auto i = 0u; i < path.size() - 1; ++i) {
-        auto nodeID = path[i]->getNodeID();
-        if (set.contains(nodeID)) {
+bool PathsOutputWriter::isAcyclic(const std::vector<ParentList*>& path,
```

Comment:
Ditto. I would try and benchmark this against an implementation that uses unordered or some other fast map implementations.

Reply:
Ditto